src/emailcontactlistmodel.cpp

```diff
@@ -0,0 +1,64 @@
+#include "emailcontactlistmodel.h"
+#include "emailcontactstorage.h"
+
+
+EmailContactListModel::EmailContactListModel(QAbstractItemModel *parent)
+    : QAbstractListModel(parent)
+{    
+}
+
+
+QHash<int, QByteArray> EmailContactListModel::roleNames() const
+{
+    static QHash<int, QByteArray> names = {
+        {NameRole, "name"},
+        {EmailRole, "email"},
+        {TimeStampRole, "timestamp"}
+    };
+    return names;
+}
+
+
+int EmailContactListModel::rowCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent)
+    return EmailContactStorage::instance()->count();
+}
+
+
+QVariant EmailContactListModel::data(const QModelIndex &index, int role) const
+{
+    EmailContactStorage *storage = EmailContactStorage::instance();
```

Comment:
The contact storage seems to be used only here and couple other places in this file. As this is kind of thin wrapper, pondering if it warrants to keep these separate or should those be just combined into one class. If the contact store is a heavy object it could be one reason, but now it seems unnecessarily heavy by going through all(?) the messages. And there's nothing refreshing it. If the stuff was instantiated by need, it would be also up-to-date by free.

Reply:
all right.
This is a thin wrapper. I dont know which is best. Now you can merge. If the storage class grows, it will increase the difficulty.